bots/topological/src/main/java/org/openjdk/skara/bots/topological/TopologicalBot.java

```diff
@@ -0,0 +1,244 @@
+                    // and pushing, until no further progress can be made.
+                    progress = false;
+                    failed = false;
+                    for (var parentsIt = parents.iterator(); parentsIt.hasNext();) {
+                        String parent = parentsIt.next();
+                        try {
+                            mergeIfAhead(repo, branch, parent);
+                            progress = true;
+                            parentsIt.remove(); // avoid doing pointless merges
+                        } catch(IOException e) {
+                            log.severe("Merge with " + parent + " failed. Reverting...");
+                            repo.abortMerge();
+                            failedMerges.add(branch + " <- " + parent);
+                            failed = true;
+                        }
+                    }
+                } while(progress && failed);
+
+                if (!failedMerges.isEmpty()) {
+                    throw new IOException("There were failed merges:\n" + failedMerges);
+                }
+            }
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+        log.info("Ending topobot run");
+    }
+
+    private List<Branch> orderedBranches(Repository repo, Path depsFile) throws IOException {
+        List<Edge> deps = new ArrayList<>();
+        for (Branch branch : branches) {
+            repo.checkout(branch);
+            if (Files.exists(depsFile)) {
+                Files.lines(depsFile)
+                        .forEach(dep -> deps.add(new Edge(new Branch(dep), branch)));
+            } else {
+                deps.add(new Edge(new Branch("master"), branch));
+            }
+        }
+        return tsort(deps).stream()
+            .filter(branch -> !branch.name().equals("master"))
+            .collect(Collectors.toList());
+    }
+
+    private void mergeIfAhead(Repository repo, Branch branch, String parent) throws IOException {
+        if (log(repo, branch.name(), parent).count() != 0) {
+            var fromHash = repo.resolve(parent).orElseThrow();
+            var isFastForward = repo.isAncestor(repo.head(), fromHash);
+            repo.merge(fromHash);
+            if (!isFastForward) {
+                log.info("Merged " + parent + " into " + branch);
+                repo.commit("Automatic merge with " + parent, "topobot", "");
+            } else {
+                log.info("Fast forwarded " + branch + " to " + parent);
+            }
+            log.info("merge with " + parent + " succeeded. The following commits will be pushed:\n"
+                    + log(repo, "origin/" + branch.name(), branch.name()).collect(Collectors.joining("\n", "\n", "\n")));
+            try {
+                repo.push(repo.head(), hostedRepo.getUrl(), branch.name());
+            } catch (IOException e) {
+                log.severe("Pusing failed! Aborting...");
+                repo.abortMerge();
```

Comment:
Hmm, don't think this `abortMerge` will actually do anything, since the working tree is clean...

Reply:
If this translates to "git merge --abort" then this is required, I think, as it will wipe out the repo of any by-product of the (failed) merge attempt (e.g. reset all files to original head).